File: jvCPP/1012.cpp
// Palindorme extendido
// https://jv.umsa.bo/problem.php?id=1012
#include <iostream>
#include <string>
using namespace std;

int main() {
    string palabra;
    while (true) {
        cin >> palabra;
        int longitud = palabra.length();
        if (longitud >= 3 && palabra.substr(0, 3) == "END")
            break;
        string palabra_invertida = string(palabra.rbegin(), palabra.rend());
        int *longitudes_prefijo = new int[longitud]();
        int k = 0;
        for (int i = 1; i < longitud; ++i) {
            k = longitudes_prefijo[i - 1];
            while (k > 0 && palabra_invertida[k] != palabra_invertida[i])
                k = longitudes_prefijo[k - 1];
            if (palabra_invertida[k] == palabra_invertida[i])
                ++k;
            longitudes_prefijo[i] = k;
        }
        k = 0;
        int i = 0;
        while (i < longitud) {
            while (k > 0 && palabra_invertida[k] != palabra[i])
                k = longitudes_prefijo[k - 1];
            if (palabra_invertida[k] == palabra[i])
                ++k;
            if (i != longitud - 1 && k == longitud)
                k = longitudes_prefijo[longitud - 1];
            ++i;
        }
        cout << palabra;
        for (int i = longitud - k - 1; i >= 0; --i)
            cout << palabra[i];
        cout << endl;
        delete[] longitudes_prefijo;
    }
    return 0;
}

File: jvPY/1012.py
# Palindrome extendido
# https://jv.umsa.bo/problem.php?id=1012



# def es_palindromo(s):
#     return s == s[::-1]

# def obtener_palindromo(string):
#     if es_palindromo(string):
#         return string

#     for i in range(len(string)):
#         subcadena = string[i:]
#         if es_palindromo(subcadena):
#             return string + string[:i][::-1]
# caso = input()
# while caso != "END":
#     resultado = obtener_palindromo(caso)
#     print(resultado)
#     caso = input()

while True:
    palabra = input()
    longitud = len(palabra)
    if palabra == 'END':
        break
    palabra_invertida = palabra[::-1]
    longitudes_prefijo = [0] * longitud
    k = 0
    for i in range(1, longitud):
        k = longitudes_prefijo[i - 1]
        while k > 0 and palabra_invertida[k] != palabra_invertida[i]:
            k = longitudes_prefijo[k - 1]
        if palabra_invertida[k] == palabra_invertida[i]:
            k += 1
        longitudes_prefijo[i] = k
    k = 0
    i = 0
    while i < longitud:
        while k > 0 and palabra_invertida[k] != palabra[i]:
            k = longitudes_prefijo[k - 1]
        if palabra_invertida[k] == palabra[i]:
            k += 1
        if i != longitud - 1 and k == longitud:
            k = longitudes_prefijo[longitud - 1]
        i += 1
    print(palabra, end='')
    for i in range(longitud - k - 1, -1, -1):
        print(palabra[i], end='')
    print()
